### PR TITLE
Improve optimize sync conditions

### PR DESCRIPTION
**Pull request by quba42**
This PR achieves the following:

- Do not disable optimize if we are switching from mirror to not mirror between two syncs. We can optmize in this case.
- Log additional reasons why we did not optimize in certain cases.
- Improve the readability of optimize conditions.

### CODE REVIEW

**Review by hstct — COMMENTED** [review]
This should be a test case as well. To test it we need to allow passing arguments to the `deb_sync_repository` fixture and adapt the `deb_init_and_sync` fixture to use the arguments as well. The actual test case should be relatively simple then.

**Comment by quba42** [comment]
@hstct I wonder if there is some simple way to assert that certain log statements have or have not happened.

**Comment by hstct** [comment]
Checking for specific log statements is somewhat messy. We can just check the task if content has been skipped or not.

**Comment by hstct** [comment]
In `test_sync.py` is a helper function that checks this provided a task and the correct report code.

Here we check if the release file and the package indices have not been skipped in the sync: https://github.com/pulp/pulp_deb/blob/main/pulp_deb/tests/functional/api/test_sync.py#L226-L227

File: pulp_deb/tests/functional/api/test_sync.py [hunk 1]
```diff
@@ -303,6 +303,27 @@ def test_sync_optimize_skip_unchanged_package_index(
...
+@pytest.mark.parallel
+def test_sync_optimize_switch_to_no_mirror(
+    deb_init_and_sync,
+):
+    """
+    Test that when syncing a repo with mirror=True, and then re-syncing that repo with
+    mirror=False, optimize=True, the releases will be skipped by optimize mode.
+    """
+
+    sync_args = {"mirror": True}
```
Comment (hstct) [inline pulp_deb/tests/functional/api/test_sync.py:315]: If we set the value here explicitly to `sync_args = {"optimize": False, "mirror": True}` the tests will run locally.

Otherwise it will fail. Here is the full log of the error:

```
c1149a7f71f9 pulp [1d43f02f-8e4a-4424-be4f-5c709066384b]: pulpcore.tasking.tasks:INFO: Starting task 018adab3-a8f9-706f-92f3-97d12a13348b
c1149a7f71f9 pulp [1d43f02f-8e4a-4424-be4f-5c709066384b]: pulp_deb.app.tasks.synchronizing:INFO: Setting optimize=False since the remote options have changed.
c1149a7f71f9 pulp [1d43f02f-8e4a-4424-be4f-5c709066384b]: pulpcore.tasking.tasks:INFO: Task 018adab3-a8f9-706f-92f3-97d12a13348b failed ('mirror')
c1149a7f71f9 pulp [1d43f02f-8e4a-4424-be4f-5c709066384b]: pulpcore.tasking.tasks:INFO:   File "/src/pulpcore/pulpcore/tasking/tasks.py", line 64, in _execute_task
c1149a7f71f9     result = func(*args, **kwargs)
c1149a7f71f9
c1149a7f71f9   File "/src/pulp_deb/pulp_deb/app/tasks/synchronizing.py", line 182, in synchronize
c1149a7f71f9     first_stage = DebFirstStage(remote, optimize, mirror, previous_repo_version)
c1149a7f71f9
c1149a7f71f9   File "/src/pulp_deb/pulp_deb/app/tasks/synchronizing.py", line 555, in __init__
c1149a7f71f9     if mirror and not previous_sync_info["sync_options"]["mirror"]:
c1149a7f71f9
```

File: pulp_deb/tests/functional/conftest.py [hunk 2]
```diff
@@ -580,7 +582,13 @@ def deb_init_and_sync(
     """Initialize a new repository and remote and sync the content from the passed URL."""
 
     def _deb_init_and_sync(
-        repository=None, remote=None, url=None, remote_args={}, repo_args={}, return_task=False
+        repository=None,
+        remote=None,
+        url=None,
+        remote_args={},
+        repo_args={},
+        sync_args={},
```
Comment (hstct) [inline pulp_deb/tests/functional/conftest.py:590]: Nitpick: this new parameter should be included in the docstring below as well.

File: pulp_deb/tests/functional/conftest.py [hunk 3]
```diff
@@ -447,14 +447,16 @@ def deb_sync_repository(apt_repository_api, monitor_task):
     and returns the monitored task.
     """
 
-    def _deb_sync_repository(remote, repo):
+    def _deb_sync_repository(remote, repo, mirror=False, optimize=True):
```
Comment (hstct) [inline pulp_deb/tests/functional/conftest.py:450]: Nitpick: The new parameters should be included in the docstring below as well.

File: pulp_deb/app/tasks/synchronizing.py [hunk 4]
```diff
@@ -546,19 +540,27 @@ def __init__(self, remote, optimize, mirror, previous_repo_version, *args, **kwa
...
+                self.optimize = False
+            if mirror and not previous_sync_info["sync_options"]["mirror"]:
+                log.info(_("Setting optimize=False since this sync switches to mirror=True."))
+                self.optimize = False
+            # TODO: https://github.com/pulp/pulp_deb/issues/631
+            if mirror:
+                log.info(_("Falling back to optimize=False behaviour since mirror=True is set!"))
+                log.info(_("See https://github.com/pulp/pulp_deb/issues/631 for more information."))
+                self.optimize = False
+                self.sync_info["sync_options"]["optimize"] = False
```
Comment (hstct) [inline pulp_deb/app/tasks/synchronizing.py:563]: In the current implementation this will likely fail in the edge case that it is the first sync of the repository. See above.

File: pulp_deb/app/tasks/synchronizing.py [hunk 5]
```diff
@@ -546,19 +540,27 @@ def __init__(self, remote, optimize, mirror, previous_repo_version, *args, **kwa
...
-            self.previous_sync_info["remote_options"] == self.sync_info["remote_options"]
-            and self.previous_sync_info["sync_options"]["mirror"]
-            == self.sync_info["sync_options"]["mirror"]
-        )
+        if self.optimize:
+            previous_sync_info = defaultdict(dict, self.previous_repo_version.info)
+            if not previous_sync_info["remote_options"] == self.sync_info["remote_options"]:
+                log.info(_("Setting optimize=False since the remote options have changed."))
+                self.optimize = False
+            if mirror and not previous_sync_info["sync_options"]["mirror"]:
```
Comment (hstct) [inline pulp_deb/app/tasks/synchronizing.py:555]: There is an edge case when a repository is synced for the first time and it is set to `mirror=True` because the `previous_sync_info` is empty and cannot resolve the nested dict condition without error.

One way to resolve this is to explicitly set the `sync_options` before in this case. Something like:

```python
if "sync_options" not in previous_sync_info:
    previous_sync_info["sync_options"] = {"mirror": None, "optimize": None}
```

Another solution would be nesting the if statements to check for this case.

File: pulp_deb/tests/functional/api/test_sync.py (hunk 1 above)
Comment (hstct) [reply pulp_deb/tests/functional/api/test_sync.py:315]: On another look this test as it is should stay because it showed that there is an edge case that will run into an error if a repository is synced for the first time and has the mirror and optimize option enabled. See below.

File: pulp_deb/app/tasks/synchronizing.py (hunk 4 above)
Comment (hstct) [inline pulp_deb/app/tasks/synchronizing.py:563]: This will likely run into the same edge case as above if we manage to get here on the repositories first sync.

File: pulp_deb/app/tasks/synchronizing.py (hunk 5 above)
Comment (quba42) [reply pulp_deb/app/tasks/synchronizing.py:555]: What I don't understand if is why this was not a problem before? We had the same dict access before...

File: pulp_deb/tests/functional/conftest.py (hunk 3 above)
Comment (quba42) [reply pulp_deb/tests/functional/conftest.py:450]: Added.

File: pulp_deb/tests/functional/conftest.py (hunk 2 above)
Comment (quba42) [reply pulp_deb/tests/functional/conftest.py:592]: Added.

File: pulp_deb/tests/functional/api/test_sync.py (hunk 1 above)
Comment (quba42) [reply pulp_deb/tests/functional/api/test_sync.py:315]: Agreed: The test caught a real bug in my changes. I was able to reproduce the bug via manual syncing using my branch.

File: pulp_deb/app/tasks/synchronizing.py (hunk 4 above)
Comment (quba42) [reply pulp_deb/app/tasks/synchronizing.py:566]: This one should be fine, since this is `self.sync_info` and not `self.previous_sync_info` and `self.sync_info` was explicitly seeded with the field a few lines further up.

File: pulp_deb/app/tasks/synchronizing.py (hunk 5 above)
Comment (quba42) [reply pulp_deb/app/tasks/synchronizing.py]: I now understood why this error was not possible before. It was only because we had the dangerous dict access hidden behind a `and` condition where the thing before the `and` would always evaluate `False` on the initial sync, thus preventing the error.

I have amended my changes to fix the issue on the PR version as well.

**Review by hstct — APPROVED** [review]
